storm-core/src/jvm/backtype/storm/messaging/netty/Client.java

```diff
@@ -42,344 +42,577 @@
                 messageBatch = null;
-                flushRequest(channel, toBeFlushed);
-                
-            } else {
-                // when channel is NOT writable, it means the internal netty buffer is full. 
-                // In this case, we can try to buffer up more incoming messages.
-                flushCheckTimer.set(System.currentTimeMillis() + flushCheckInterval);
+                flushMessages(channel, toBeFlushed);
+            }
+            else {
+                // We cannot write to the channel, which means Netty's internal write buffer is full.
+                // In this case, we buffer the remaining messages and wait for the next messages to arrive.
+                //
+                // Background:
+                // Netty 3.x maintains an internal write buffer with a high water mark for each channel (default: 64K).
+                // This represents the amount of data waiting to be flushed to operating system buffers.  If the
+                // outstanding data exceeds this value then the channel is set to non-writable.  When this happens, a
+                // INTEREST_CHANGED channel event is triggered.  Netty sets the channel to writable again once the data
+                // has been flushed to the system buffers.
+                //
+                // See http://stackoverflow.com/questions/14049260
+                resumeBackgroundFlushing();
+                nextBackgroundFlushTimeMs.set(nowMillis() + flushCheckIntervalMs);
             }
         }
 
     }
 
-    public String name() {
-        if (null != remote_addr) {
-            return PREFIX + remote_addr.toString();
-        }
-        return "";
+    private boolean hasMessages(Iterator<TaskMessage> msgs) {
+        return msgs != null && msgs.hasNext();
     }
 
-    private synchronized void flush(Channel channel) {
-        if (!closing) {
-            if (null != messageBatch && !messageBatch.isEmpty()) {
-                MessageBatch toBeFlushed = messageBatch;
-                flushCheckTimer.set(Long.MAX_VALUE);
-                flushRequest(channel, toBeFlushed);
-                messageBatch = null;
+    /**
+     * We will drop pending messages and let at-least-once message replay kick in.
+     *
+     * Another option would be to buffer the messages in memory.  But this option has the risk of causing OOM errors,
+     * especially for topologies that disable message acking because we don't know whether the connection recovery will
+     * succeed  or not, and how long the recovery will take.
+     */
+    private void handleMessagesWhenConnectionIsUnavailable(Iterator<TaskMessage> msgs) {
+        LOG.error("connection to {} is unavailable", dstAddressPrefixedName);
+        dropPendingMessages(msgs);
+    }
+
+    private void dropPendingMessages(Iterator<TaskMessage> msgs) {
+        // We consume the iterator by traversing and thus "emptying" it.
+        int msgCount = iteratorSize(msgs);
```

Comment:
We are throwing away messages, it would be nice to include that in the state so we can know how many messages we threw away waiting for a worker to come up.